### PR TITLE
feat: Added bootstrapped ci to plots

### PR DESCRIPTION
**Pull request by KennethEnevoldsen**
- [x] I have considered whether this PR needs review

Added bootstrapped CI to sex and age. Also made it easy to add it to almost any other plots. It currently would look something like this:

Screenshot 2023-04-05 at 14 47 39

### CODE REVIEW

**Comment by KennethEnevoldsen** [comment]
@MartinBernstorff, it seems like the coverage is problematic here. I would love to get information about the coverage in notably to see if new lines are covered.

**Comment by KennethEnevoldsen** [comment]
@MartinBernstorff 

PS: the stalebot is maybe a bit too aggressive

**Comment by MartinBernstorff** [comment]
> @MartinBernstorff
> 
> PS: the stalebot is maybe a bit too aggressive

Agreed, I raised it here: https://github.com/MartinBernstorff/swift-python-cookiecutter/issues/9

**Comment by MartinBernstorff** [comment]
> @MartinBernstorff, it seems like the coverage is problematic here. I would love to get information about the coverage in notably to see if new lines are covered.

Completely agreed. Added an issue here: https://github.com/MartinBernstorff/swift-python-cookiecutter/issues/99

File: src/psycop_model_evaluation/binary/subgroups/base.py [hunk 1]
```diff
@@ -37,16 +41,20 @@ def create_roc_auc_by_input(
     )
 
     # bin data and calculate metric per bin
+    calc_perf = partial(
+        calc_performance,
+        metric=roc_auc_score,
+        confidence_interval=confidence_interval,
+    )
```
Comment (MartinBernstorff) [inline src/psycop_model_evaluation/binary/subgroups/base.py:48]: Nice little refactor 👍

File: src/psycop_model_evaluation/binary/utils.py [hunk 2]
```diff
@@ -30,6 +41,35 @@ def calc_performance(df: pd.DataFrame, metric: Callable) -> pd.Series:
...
+        _kwargs.update(kwargs)
+
+        # Calculate the confidence interval
+        def metric_wrapper(
+            true: np.ndarray,
+            pred: np.ndarray,
+            **kwargs: Any,  # noqa: ARG001
+        ) -> float:
+            # bootstrap function requires the metric function to
+            # be able to additional arguments (notably the length of the array)
```
Comment (MartinBernstorff) [inline src/psycop_model_evaluation/binary/utils.py:59]:
```suggestion
            # be able to take additional arguments (notably the length of the array)
```

File: src/psycop_model_evaluation/binary/utils.py [hunk 3]
```diff
@@ -30,6 +41,35 @@ def calc_performance(df: pd.DataFrame, metric: Callable) -> pd.Series:
...
+            return metric(true, pred)
+
+        boot = bootstrap(
+            (df["y"], df["y_hat"]),
+            statistic=metric_wrapper,
+            confidence_level=confidence_interval,
+            paired=True,
```
Comment (MartinBernstorff) [inline src/psycop_model_evaluation/binary/utils.py:66]: Yeah, most of our observations are probably paired, so happy with this choice 👍

**Review by MartinBernstorff — COMMENTED** [review]
Looks great, thanks a lot for taking a look at this! ❤️

I'm not sure how well this implementation transfers to:
* roc_auc 
* precision_recall
* timedelta_data

Since we're touching `calc_performance`, could you check those before we merge? :-)

**Comment by KennethEnevoldsen** [comment]
calc_performance is backwards compatible as it stands so there shouldn't be any problem

File: src/psycop_model_evaluation/binary/utils.py (hunk 3 above)
Comment (MartinBernstorff) [reply src/psycop_model_evaluation/binary/utils.py:66]: Ah lol, thanks!

**Comment by MartinBernstorff** [comment]
> calc_performance is backwards compatible as it stands so there shouldn't be any problem (i.e. unless you add the confidence interval or kwargs argument nothing have changed).

Oh, fair point! I wonder whether this solves the entire problem, though? If we have a bunch of plots this can't be applied to, we might want to look for an alternate solution. Since you have the code fresh in your mind, I imagine you could test it _way_ faster than I could 🚀❤️

**Comment by KennethEnevoldsen** [comment]
> Oh, fair point! I wonder whether this solves the entire problem, though? If we have a bunch of plots this can't be applied to, we might want to look for an alternate solution. Since you have the code fresh in your mind, I imagine you could test it way faster than I could 🚀

Seems like it was quite easy to add to some places, but some would need a bit extra - notably the cases where the base_chart is *overloaded* to also handle multiple lines (I would actually refactor is such that base_chart only ever plots 1 thing and then it would be possible to call it multiple times, but out of scope for this PR).

**Comment by MartinBernstorff** [comment]
That's what I expected, thanks a lot for taking a look at it! Will resolve conflicts and merge.

**Pull request merged**